checker/src/org/checkerframework/checker/lock/wholefiles/LockAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,261 @@
+            super(f, LOCKHELD);
+        }
+
+        private boolean isGuardedBy(AnnotationMirror am) {
+            return AnnotationUtils.areSameIgnoringValues(am, GUARDEDBY) ||
+                   AnnotationUtils.areSameIgnoringValues(am, JAVAXGUARDEDBY) ||
+                   AnnotationUtils.areSameIgnoringValues(am, JCIPGUARDEDBY);
+        }
+
+        @Override
+        public boolean isSubtype(AnnotationMirror rhs, AnnotationMirror lhs) {
+
+            boolean lhsIsGuardedBy = isGuardedBy(lhs);
+            boolean rhsIsGuardedBy = isGuardedBy(rhs);
+
+            if (lhsIsGuardedBy && rhsIsGuardedBy) {
+                // Two @GuardedBy annotations are considered subtypes of each other if and only if their values match exactly.
+
+                List<String> lhsValues =
+                    AnnotationUtils.getElementValueArray(lhs, "value", String.class, true);
+                List<String> rhsValues =
+                    AnnotationUtils.getElementValueArray(rhs, "value", String.class, true);
+
+                return rhsValues.containsAll(lhsValues) && lhsValues.containsAll(rhsValues);
+            }
+
+            // Remove values from @GuardedBy annotations (and use the Checker Framework's GuardedBy annotation, not JCIP's or Javax's)
+            // for further subtype checking.
+
+            return super.isSubtype(rhsIsGuardedBy ? GUARDEDBY : rhs, lhsIsGuardedBy ? GUARDEDBY : lhs);
+        }
+    }
+
+    // Returns the list of expressions specified in the @Holding (or equivalent) annotation on a method.
+    // Returns an empty list if such an annotation is not present.
+    protected List<String> methodHolding(ExecutableElement element) {
+        AnnotationMirror holding = getDeclAnnotation(element, checkerHoldingClass);
+        AnnotationMirror guardedBy
+            = getDeclAnnotation(element, jcipGuardedByClass);
+        AnnotationMirror guardedByJavax
+            = getDeclAnnotation(element, javaxGuardedByClass);
+
+        if (holding == null && guardedBy == null && guardedByJavax == null)
+            return Collections.emptyList();
+
+        List<String> locks = new ArrayList<String>();
+
+        if (holding != null) {
+            List<String> holdingValue = AnnotationUtils.getElementValueArray(holding, "value", String.class, false);
+            locks.addAll(holdingValue);
+        }
+        if (guardedBy != null) {
+            String guardedByValue = AnnotationUtils.getElementValue(guardedBy, "value", String.class, false);
+            locks.add(guardedByValue);
+        }
+        if (guardedByJavax != null) {
+            String guardedByValue = AnnotationUtils.getElementValue(guardedByJavax, "value", String.class, false);
+            locks.add(guardedByValue);
+        }
+
+        return locks;
+    }
+
+    // The side effect annotations processed by the Lock Checker,
+    // in order of increasingly strong guarantees.
+    enum SideEffectAnnotation {
```

Comment:
Seems bad to encode the "strongness" of these annotations by order in the enum.  At the every least, I recommend adding an isWeaker method (or isStronger) and replacing code that compares ordinals with a call to that method.